File: src/main/java/pers/neige/neigeitems/action/node/impl/NumberParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.utils.ListUtils;
import pers.neige.neigeitems.utils.ScriptUtils;
import pers.neige.neigeitems.utils.StringUtils;

import java.math.BigDecimal;
import java.util.concurrent.ThreadLocalRandom;

public class NumberParser extends NodeParser {
    public NumberParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "number";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            getParsedValue(context, params, "min"),
            getParsedValue(context, params, "max"),
            getParsedValue(context, params, "fixed"),
            getParsedValue(context, params, "mode")
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val paramsList = StringUtils.split(params, '_', '\\', 4);
        return handle(
            paramsList.get(0),
            ListUtils.getOrNull(paramsList, 1),
            ListUtils.getOrNull(paramsList, 2),
            ListUtils.getOrNull(paramsList, 3)
        );
    }

    private @Nullable String handle(
        @Nullable String min,
        @Nullable String max,
        @Nullable String fixed,
        @Nullable String mode
    ) {
        if (min == null) {
            warning("未指定最小值");
            return null;
        }

        if (max == null) {
            warning("未指定最大值");
            return null;
        }

        val minValue = parse(min, 0D, pers.neige.neigeitems.utils.NumberParser::parseDouble, "{0} 并非数字, 无法用作最小值");
        if (minValue == null) return null;

        val maxValue = parse(max, 0D, pers.neige.neigeitems.utils.NumberParser::parseDouble, "{0} 并非数字, 无法用作最大值");
        if (maxValue == null) return null;

        val result = minValue >= maxValue ? minValue : ThreadLocalRandom.current().nextDouble(minValue, maxValue);

        val fixedValue = parse(fixed, 0, pers.neige.neigeitems.utils.NumberParser::parseInteger, "{0} 并非数字, 无法用作取整位数");
        if (fixedValue == null) return null;

        val roundingMode = ScriptUtils.toRoundingMode(mode);

        return new BigDecimal(Double.toString(result)).setScale(fixedValue, roundingMode).toString();
    }
}

File: project/common/src/main/kotlin/pers/neige/neigeitems/section/impl/NumberParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.ScriptUtils.toRoundingMode
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import java.util.concurrent.ThreadLocalRandom

/**
 * 随机数节点解析器
 */
object NumberParser : SectionParser() {
    override val id: String = "number"

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String? {
        return handler(
            cache,
            player,
            sections,
            true,
            data.getString("min"),
            data.getString("max"),
            data.getString("fixed"),
            data.getString("mode")
        )
    }

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(
            cache,
            player,
            sections,
            false,
            args.getOrNull(0),
            args.getOrNull(1),
            args.getOrNull(2),
            args.getOrNull(3)
        ) ?: "<$id::${args.joinToString("_")}>"
    }

    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param parse 是否对参数进行节点解析
     * @param minString 最小值文本
     * @param maxString 最大值文本
     * @param fixedString 取整位数文本
     * @param roundingMode 取整模式
     * @return 解析值
     */
    private fun handler(
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        parse: Boolean,
        minString: String?,
        maxString: String?,
        fixedString: String?,
        roundingMode: String?
    ): String? {
        // 获取大小范围
        val min = minString?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        val max = maxString?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        // 获取取整位数
        val fixed = fixedString?.parseSection(parse, cache, player, sections)?.toIntOrNull() ?: 0
        // 获取取整模式
        val mode = roundingMode.toRoundingMode()
        // 获取随机数
        if (min != null && max != null) {
            val num =
                if (min >= max)
                    min
                else
                    ThreadLocalRandom.current().nextDouble(min, max)
            return num.toBigDecimal().setScale(fixed, mode).toString()
        }
        return null
    }
}